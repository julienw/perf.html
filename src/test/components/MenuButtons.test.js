/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import MenuButtons from '../../components/app/MenuButtons';
import { MenuButtonsMetaInfo } from '../../components/app/MenuButtons/MetaInfo';
import { render, fireEvent, wait } from 'react-testing-library';
import { Provider } from 'react-redux';
import { storeWithProfile } from '../fixtures/stores';
import { TextEncoder } from 'util';
import { stateFromLocation } from '../../app-logic/url-handling';
import { ensureExists } from '../../utils/flow';
import {
  getProfileFromTextSamples,
  getProfileWithMarkers,
} from '../fixtures/profiles/processed-profile';
import { createGeckoProfile } from '../fixtures/profiles/gecko-profile';
import { processProfile } from '../../profile-logic/process-profile';
import type { Profile, SymbolicationStatus } from 'firefox-profiler/types';

// We mock profile-store but we want the real error, so that we can simulate it.
import { uploadBinaryProfileData } from '../../profile-logic/profile-store';
jest.mock('../../profile-logic/profile-store');
const { UploadAbortedError } = jest.requireActual(
  '../../profile-logic/profile-store'
);

// Mocking sha1
import sha1 from '../../utils/sha1';
jest.mock('../../utils/sha1');

// Mocking compress
jest.mock('../../utils/gz');

// Mocking shortenUrl
import { shortenUrl } from '../../utils/shorten-url';
jest.mock('../../utils/shorten-url');

// Mock hash
const hash = 'c5e53f9ab6aecef926d4be68c84f2de550e2ac2f';

describe('app/MenuButtons', function() {
  function mockUpload() {
    // Create a promise with the resolve function outside of it.
    let resolveUpload, rejectUpload;
    const promise = new Promise((resolve, reject) => {
      resolveUpload = resolve;
      rejectUpload = reject;
    });

    promise.catch(() => {
      // Node complains if we don't handle a promise/catch, and this one may reject
      // before it's properly handled. Catch it here so that Node doesn't complain.
    });

    // Flow doesn't know uploadBinaryProfileData is a jest mock.
    (uploadBinaryProfileData: any).mockImplementation(
      (() => ({
        abortFunction: () => {
          // In the real implementation, we call xhr.abort, hwich in turn
          // triggers an "abort" event on the XHR object, which in turn rejects
          // the promise with the error UploadAbortedError. So we do just that
          // here directly, to simulate this.
          rejectUpload(new UploadAbortedError());
        },
        startUpload: () => promise,
      }): typeof uploadBinaryProfileData)
    );

    return { resolveUpload, rejectUpload };
  }

  function createSimpleProfile(updateChannel = 'release') {
    const { profile } = getProfileFromTextSamples('A');
    profile.meta.updateChannel = updateChannel;
    return { profile };
  }

  function createPreferenceReadProfile(updateChannel = 'release') {
    const profile = getProfileWithMarkers([
      [
        'PreferenceRead',
        1,
        {
          type: 'PreferenceRead',
          startTime: 0,
          endTime: 1,
          prefAccessTime: 0,
          prefName: 'testing',
          prefKind: 'testing',
          prefType: 'testing',
          prefValue: 'testing',
        },
      ],
    ]);
    profile.meta.updateChannel = updateChannel;
    return { profile };
  }

  function setup(profile) {
    jest.useFakeTimers();

    const store = storeWithProfile(profile);
    const { resolveUpload, rejectUpload } = mockUpload();

    store.dispatch({
      type: 'UPDATE_URL_STATE',
      newUrlState: stateFromLocation({
        pathname: '/from-addon',
        search: '',
        hash: '',
      }),
    });

    const renderResult = render(
      <Provider store={store}>
        <MenuButtons />
      </Provider>
    );

    const { container, getByTestId, getByText, queryByText } = renderResult;
    const getPublishButton = () => getByText('Publish…');
    const getErrorButton = () => getByText('Error publishing…');
    const getCancelButton = () => getByText('Cancel Upload');
    const getPanelForm = () =>
      ensureExists(
        container.querySelector('form'),
        'Could not find the form in the panel'
      );
    const queryPreferenceCheckbox = () =>
      queryByText('Include preference values');
    const getPanel = () => getByTestId('MenuButtonsPublish-container');
    const clickAndRunTimers = where => {
      fireEvent.click(where);
      jest.runAllTimers();
    };

    return {
      store,
      ...renderResult,
      getPanel,
      getPublishButton,
      getErrorButton,
      getCancelButton,
      getPanelForm,
      queryPreferenceCheckbox,
      clickAndRunTimers,
      resolveUpload,
      rejectUpload,
    };
  }

  describe('<Publish>', function() {
    beforeAll(function() {
      if ((window: any).TextEncoder) {
        throw new Error('A TextEncoder was already on the window object.');
      }
      (window: any).TextEncoder = TextEncoder;
    });

    afterAll(async function() {
      delete URL.createObjectURL;
      delete URL.revokeObjectURL;
      delete (window: any).TextEncoder;
    });

    beforeEach(function() {
      // Flow doesn't know sha1 is a jest mock.
      (sha1: any).mockImplementation((_data: Uint8Array) =>
        Promise.resolve(hash)
      );
      // Flow doesn't know shortenUrl is a jest mock.
      (shortenUrl: any).mockImplementation(() =>
        Promise.resolve('https://profiler.firefox.com/')
      );
      // jsdom does not have URL.createObjectURL.
      // See https://github.com/jsdom/jsdom/issues/1721
      (URL: any).createObjectURL = () => 'mockCreateObjectUrl';
      (URL: any).revokeObjectURL = () => {};
    });

    it('matches the snapshot for the closed state', () => {
      const { profile } = createSimpleProfile();
      const { container } = setup(profile);
      expect(container).toMatchSnapshot();
    });

    it('matches the snapshot for the opened panel for a nightly profile', () => {
      const { profile } = createSimpleProfile('nightly');
      const { getPanel, getPublishButton, clickAndRunTimers } = setup(profile);
      clickAndRunTimers(getPublishButton());
      expect(getPanel()).toMatchSnapshot();
    });

    it('matches the snapshot for the opened panel for a release profile', () => {
      const { profile } = createSimpleProfile('release');
      const { getPanel, getPublishButton, clickAndRunTimers } = setup(profile);
      clickAndRunTimers(getPublishButton());
      expect(getPanel()).toMatchSnapshot();
    });

    it('shows the Include preference values checkbox when a PreferenceRead marker is in the profile', () => {
      const { profile } = createPreferenceReadProfile('release');
      const {
        getPublishButton,
        clickAndRunTimers,
        queryPreferenceCheckbox,
      } = setup(profile);
      clickAndRunTimers(getPublishButton());
      expect(queryPreferenceCheckbox()).toBeTruthy();
    });

    it('does not show the Include preference values checkbox when a PreferenceRead marker is in the profile', () => {
      const { profile } = createSimpleProfile('release');
      const {
        getPublishButton,
        clickAndRunTimers,
        queryPreferenceCheckbox,
      } = setup(profile);
      clickAndRunTimers(getPublishButton());
      expect(queryPreferenceCheckbox()).toBeFalsy();
    });

    it('can publish, cancel, and then publish again', async () => {
      const { profile } = createSimpleProfile();
      const {
        getPanel,
        getPublishButton,
        getCancelButton,
        getPanelForm,
        clickAndRunTimers,
      } = setup(profile);
      clickAndRunTimers(getPublishButton());
      fireEvent.submit(getPanelForm());

      // These shouldn't exist anymore.
      expect(() => getPanel()).toThrow();
      expect(() => getPublishButton()).toThrow();

      clickAndRunTimers(getCancelButton());

      // This might be asynchronous, depending on the underlying code.
      await wait(() => {
        getPublishButton();
      });
      expect(getPublishButton()).toBeTruthy();
    });

    it('matches the snapshot for an error', async () => {
      const { profile } = createSimpleProfile();
      const {
        getPanel,
        getPublishButton,
        getErrorButton,
        getPanelForm,
        rejectUpload,
        clickAndRunTimers,
      } = setup(profile);

      clickAndRunTimers(getPublishButton());
      fireEvent.submit(getPanelForm());
      rejectUpload('This is a mock error');

      // Wait until the error button is visible.
      await wait(() => {
        getErrorButton();
      });

      // Now click the error button, and get a snapshot of the panel.
      clickAndRunTimers(getErrorButton());
      expect(getPanel()).toMatchSnapshot();
    });
  });
});

describe('<MenuButtonsMetaInfo>', function() {
  function setup(profile: Profile, symbolicationStatus = 'DONE') {
    jest.useFakeTimers();
    jest.spyOn(Date.prototype, 'toLocaleString').mockImplementation(function() {
      // eslint-disable-next-line babel/no-invalid-this
      return 'toLocaleString ' + this.toUTCString();
    });
    const store = storeWithProfile(profile);
    const resymbolicateProfile = jest.fn();

    const renderResults = render(
      <Provider store={store}>
        <MenuButtonsMetaInfo
          profile={profile}
          resymbolicateProfile={resymbolicateProfile}
          symbolicationStatus={symbolicationStatus}
        />
      </Provider>
    );

    return {
      store,
      resymbolicateProfile,
      renderResults,
      ...renderResults,
    };
  }

  it('matches the snapshot', async () => {
    // Using gecko profile because it has metadata and profilerOverhead data in it.
    const profile = processProfile(createGeckoProfile());
    profile.meta.configuration = {
      features: ['js', 'threads'],
      threads: ['GeckoMain', 'DOM Worker'],
      capacity: Math.pow(2, 14),
      duration: 20,
    };

    const { container, getByValue } = setup(profile);
    const metaInfoButton = getByValue('Firefox 48 – macOS 10.11');
    fireEvent.click(metaInfoButton);
    jest.runAllTimers();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('with no statistics object should not make the app crash', async () => {
    // Using gecko profile because it has metadata and profilerOverhead data in it.
    const profile = processProfile(createGeckoProfile());
    // We are removing statistics objects from all overhead objects to test
    // the robustness of our handling code.
    if (profile.profilerOverhead) {
      for (const overhead of profile.profilerOverhead) {
        delete overhead.statistics;
      }
    }

    const { getByValue, container } = setup(profile);

    const metaInfoButton = getByValue('Firefox 48 – macOS 10.11');
    fireEvent.click(metaInfoButton);
    jest.runAllTimers();

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('symbolication', function() {
    type SymbolicationTestConfig = {|
      symbolicated: boolean,
      symbolicationStatus: SymbolicationStatus,
    |};

    function setupSymbolicationTest(config: SymbolicationTestConfig) {
      const { profile } = getProfileFromTextSamples('A');
      profile.meta.symbolicated = config.symbolicated;

      const setupResult = setup(profile, config.symbolicationStatus);

      // Open up the arrow panel for the test.
      const { getByValue } = setupResult;
      fireEvent.click(getByValue('Firefox'));
      jest.runAllTimers();

      return setupResult;
    }

    it('handles successfully symbolicated profiles', () => {
      const { getByText, resymbolicateProfile } = setupSymbolicationTest({
        symbolicated: true,
        symbolicationStatus: 'DONE',
      });

      expect(getByText('Profile is symbolicated')).toBeTruthy();
      getByText('Re-symbolicate profile').click();
      expect(resymbolicateProfile).toHaveBeenCalled();
    });

    it('handles the contradictory state of non-symbolicated profiles that are done', () => {
      const { getByText, resymbolicateProfile } = setupSymbolicationTest({
        symbolicated: false,
        symbolicationStatus: 'DONE',
      });

      expect(getByText('Profile is not symbolicated')).toBeTruthy();
      getByText('Symbolicate profile').click();
      expect(resymbolicateProfile).toHaveBeenCalled();
    });

    it('handles in progress symbolication', () => {
      const { getByText, queryByText } = setupSymbolicationTest({
        symbolicated: false,
        symbolicationStatus: 'SYMBOLICATING',
      });

      expect(getByText('Currently symbolicating profile')).toBeTruthy();
      // No symbolicate button is available.
      expect(queryByText('Symbolicate profile')).toBeFalsy();
      expect(queryByText('Re-symbolicate profile')).toBeFalsy();
    });

    it('handles in progress re-symbolication', () => {
      const { getByText, queryByText } = setupSymbolicationTest({
        symbolicated: true,
        symbolicationStatus: 'SYMBOLICATING',
      });

      expect(getByText('Attempting to re-symbolicate profile')).toBeTruthy();
      // No symbolicate button is available.
      expect(queryByText('Symbolicate profile')).toBeFalsy();
      expect(queryByText('Re-symbolicate profile')).toBeFalsy();
    });
  });
});
