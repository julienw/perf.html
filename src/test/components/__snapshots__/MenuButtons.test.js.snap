// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MetaInfoPanel> Full View Button is present when we are in the active tab view 1`] = `
<div>
  <button
    class="menuButtonsButton menuButtonsButton-hasIcon menuButtonsRevertToFullView"
    type="button"
  >
    Full View
  </button>
  <div
    class="buttonWithPanel"
  >
    <button
      aria-expanded="false"
      class="buttonWithPanelButton menuButtonsButton menuButtonsMetaInfoButtonButton menuButtonsButton-hasIcon menuButtonsMetaInfoButtonButton-local"
      type="button"
    >
      Local Profile
    </button>
  </div>
  <div
    class="buttonWithPanel"
  >
    <button
      aria-expanded="false"
      class="buttonWithPanelButton menuButtonsButton menuButtonsShareButtonButton menuButtonsButton-hasIcon"
      type="button"
    >
      Upload
    </button>
  </div>
  <a
    class="menuButtonsButton menuButtonsButton-hasLeftBorder"
    href="/docs/"
    target="_blank"
    title="Open the documentation in a new window"
  >
    Docs
    <i
      class="open-in-new"
    />
  </a>
</div>
`;

exports[`<MetaInfoPanel> matches the snapshot 1`] = `
<div>
  <h2
    class="metaInfoSubTitle"
  >
    Profile Information
  </h2>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Recording started:
      </span>
      toLocaleString Sat, 09 Apr 2016 17:02:32 GMT
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Interval:
      </span>
      1ms
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Profile Version:
      </span>
      33
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Buffer Capacity:
      </span>
      16.0KB
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Buffer Duration:
      </span>
      20 seconds
    </div>
    <div
      class="metaInfoSection"
    >
      <div
        class="metaInfoRow"
      >
        <span
          class="metaInfoLabel"
        >
          Features
          :
        </span>
        <ul
          class="metaInfoList"
        >
          <li
            class="metaInfoListItem"
          >
            js
          </li>
          <li
            class="metaInfoListItem"
          >
            threads
          </li>
        </ul>
      </div>
      <div
        class="metaInfoRow"
      >
        <span
          class="metaInfoLabel"
        >
          Threads Filter
          :
        </span>
        <ul
          class="metaInfoList"
        >
          <li
            class="metaInfoListItem"
          >
            GeckoMain
          </li>
          <li
            class="metaInfoListItem"
          >
            DOM Worker
          </li>
        </ul>
      </div>
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Symbols:
      </span>
      Profile is not symbolicated
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      />
      <button
        class="photon-button photon-button-micro"
        type="button"
      >
        Symbolicate profile
      </button>
    </div>
  </div>
  <h2
    class="metaInfoSubTitle"
  >
    Application
  </h2>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Name and version:
      </span>
      Firefox 48
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Update Channel:
      </span>
      nightly
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Build ID:
      </span>
      20181126165837
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Build Type:
      </span>
      Debug
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Extensions
        :
      </span>
      <ul
        class="metaInfoList"
      >
        <li
          class="metaInfoListItem"
        >
          Form Autofill
        </li>
        <li
          class="metaInfoListItem"
        >
          Firefox Screenshots
        </li>
        <li
          class="metaInfoListItem"
        >
          Gecko Profiler
        </li>
      </ul>
    </div>
  </div>
  <h2
    class="metaInfoSubTitle"
  >
    Platform
  </h2>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        OS:
      </span>
      macOS 10.11
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        ABI:
      </span>
      x86_64-gcc3
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        CPU:
      </span>
      4 physical cores
      , 
      8 logical cores
    </div>
  </div>
  <details>
    <summary
      class="arrowPanelSubTitle"
    >
      Profiler Overhead
    </summary>
    <div
      class="arrowPanelSection"
    >
      <div
        class="metaInfoGrid"
      >
        <div />
        <div>
          Mean
        </div>
        <div>
          Max
        </div>
        <div>
          Min
        </div>
        <div>
          Overhead
        </div>
        <div>
          227μs
        </div>
        <div>
          410μs
        </div>
        <div>
          38μs
        </div>
        <div>
          Cleaning
        </div>
        <div>
          54μs
        </div>
        <div>
          100μs
        </div>
        <div>
          7.0μs
        </div>
        <div>
          Counter
        </div>
        <div>
          59μs
        </div>
        <div>
          105μs
        </div>
        <div>
          12μs
        </div>
        <div>
          Interval
        </div>
        <div>
          857μs
        </div>
        <div>
          1,000μs
        </div>
        <div>
          0.000μs
        </div>
        <div>
          Lockings
        </div>
        <div>
          49μs
        </div>
        <div>
          95μs
        </div>
        <div>
          2.0μs
        </div>
      </div>
      <div
        class="metaInfoRow"
      >
        <span
          class="metaInfoWideLabel"
        >
          Overhead Durations:
        </span>
        <span
          class="metaInfoValueRight"
        >
          1,586μs
        </span>
      </div>
      <div
        class="metaInfoRow"
      >
        <span
          class="metaInfoWideLabel"
        >
          Overhead Percentage:
        </span>
        <span
          class="metaInfoValueRight"
        >
          26,433%
        </span>
      </div>
      <div
        class="metaInfoRow"
      >
        <span
          class="metaInfoWideLabel"
        >
          Profiled Duration:
        </span>
        <span
          class="metaInfoValueRight"
        >
          6.0μs
        </span>
      </div>
    </div>
  </details>
</div>
`;

exports[`<MetaInfoPanel> with no statistics object should not make the app crash 1`] = `
<div>
  <h2
    class="metaInfoSubTitle"
  >
    Profile Information
  </h2>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Recording started:
      </span>
      toLocaleString Sat, 09 Apr 2016 17:02:32 GMT
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Interval:
      </span>
      1ms
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Profile Version:
      </span>
      33
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Symbols:
      </span>
      Profile is not symbolicated
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      />
      <button
        class="photon-button photon-button-micro"
        type="button"
      >
        Symbolicate profile
      </button>
    </div>
  </div>
  <h2
    class="metaInfoSubTitle"
  >
    Application
  </h2>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Name and version:
      </span>
      Firefox 48
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Update Channel:
      </span>
      nightly
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Build ID:
      </span>
      20181126165837
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Build Type:
      </span>
      Debug
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        Extensions
        :
      </span>
      <ul
        class="metaInfoList"
      >
        <li
          class="metaInfoListItem"
        >
          Form Autofill
        </li>
        <li
          class="metaInfoListItem"
        >
          Firefox Screenshots
        </li>
        <li
          class="metaInfoListItem"
        >
          Gecko Profiler
        </li>
      </ul>
    </div>
  </div>
  <h2
    class="metaInfoSubTitle"
  >
    Platform
  </h2>
  <div
    class="metaInfoSection"
  >
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        OS:
      </span>
      macOS 10.11
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        ABI:
      </span>
      x86_64-gcc3
    </div>
    <div
      class="metaInfoRow"
    >
      <span
        class="metaInfoLabel"
      >
        CPU:
      </span>
      4 physical cores
      , 
      8 logical cores
    </div>
  </div>
</div>
`;

exports[`app/MenuButtons <Publish> can publish and revert 1`] = `
<body>
  <div>
    <button
      class="menuButtonsButton menuButtonsButton-hasIcon menuButtonsRevertButton"
      type="button"
    >
      Revert to Original Profile
    </button>
    <div
      class="buttonWithPanel"
    >
      <button
        aria-expanded="false"
        class="buttonWithPanelButton menuButtonsButton menuButtonsMetaInfoButtonButton menuButtonsButton-hasIcon menuButtonsMetaInfoButtonButton-uploaded"
        type="button"
      >
        Uploaded Profile
      </button>
    </div>
    <div
      class="buttonWithPanel"
    >
      <button
        aria-expanded="false"
        class="buttonWithPanelButton menuButtonsButton menuButtonsShareButtonButton menuButtonsButton-hasIcon"
        type="button"
      >
        Re-upload
      </button>
    </div>
    <div
      class="buttonWithPanel open"
    >
      <button
        aria-expanded="true"
        class="buttonWithPanelButton menuButtonsButton menuButtonsButton-hasIcon menuButtonsPermalinkButtonButton"
        type="button"
      >
        Permalink
      </button>
      <div
        class="arrowPanelAnchor"
      >
        <div
          class="arrowPanel open menuButtonsPermalinkPanel"
        >
          <div
            class="arrowPanelArrow"
          />
          <div
            class="arrowPanelContent"
          >
            <input
              class="menuButtonsPermalinkTextField photon-input"
              data-testid="MenuButtonsPermalink-input"
              readonly=""
              type="text"
              value="https://profiler.firefox.com/"
            />
          </div>
        </div>
      </div>
    </div>
    <a
      class="menuButtonsButton menuButtonsButton-hasLeftBorder"
      href="/docs/"
      target="_blank"
      title="Open the documentation in a new window"
    >
      Docs
      <i
        class="open-in-new"
      />
    </a>
  </div>
</body>
`;

exports[`app/MenuButtons <Publish> matches the snapshot for an error 1`] = `
<div
  class="menuButtonsPublishUpload"
  data-testid="MenuButtonsPublish-container"
>
  <div
    class="photon-message-bar photon-message-bar-error photon-message-bar-inner-content"
  >
    <div
      class="photon-message-bar-inner-text"
    >
      Uh oh, something went wrong when uploading the profile.
    </div>
    <button
      class="photon-button photon-button-micro photon-message-bar-action-button"
      type="button"
    >
      Try again
    </button>
  </div>
  <div
    class="menuButtonsPublishError"
  >
    This is a mock error
  </div>
</div>
`;

exports[`app/MenuButtons <Publish> matches the snapshot for the closed state 1`] = `
<div>
  <div
    class="buttonWithPanel"
  >
    <button
      aria-expanded="false"
      class="buttonWithPanelButton menuButtonsButton menuButtonsMetaInfoButtonButton menuButtonsButton-hasIcon menuButtonsMetaInfoButtonButton-local"
      type="button"
    >
      Local Profile
    </button>
  </div>
  <div
    class="buttonWithPanel"
  >
    <button
      aria-expanded="false"
      class="buttonWithPanelButton menuButtonsButton menuButtonsShareButtonButton menuButtonsButton-hasIcon"
      type="button"
    >
      Upload
    </button>
  </div>
  <a
    class="menuButtonsButton menuButtonsButton-hasLeftBorder"
    href="/docs/"
    target="_blank"
    title="Open the documentation in a new window"
  >
    Docs
    <i
      class="open-in-new"
    />
  </a>
</div>
`;

exports[`app/MenuButtons <Publish> matches the snapshot for the menu buttons and the opened panel for an already uploaded profile 1`] = `
<div>
  <div
    class="buttonWithPanel"
  >
    <button
      aria-expanded="false"
      class="buttonWithPanelButton menuButtonsButton menuButtonsMetaInfoButtonButton menuButtonsButton-hasIcon menuButtonsMetaInfoButtonButton-uploaded"
      type="button"
    >
      Uploaded Profile
    </button>
  </div>
  <div
    class="buttonWithPanel"
  >
    <button
      aria-expanded="false"
      class="buttonWithPanelButton menuButtonsButton menuButtonsShareButtonButton menuButtonsButton-hasIcon"
      type="button"
    >
      Re-upload
    </button>
  </div>
  <div
    class="buttonWithPanel"
  >
    <button
      aria-expanded="false"
      class="buttonWithPanelButton menuButtonsButton menuButtonsButton-hasIcon menuButtonsPermalinkButtonButton"
      type="button"
    >
      Permalink
    </button>
  </div>
  <a
    class="menuButtonsButton menuButtonsButton-hasLeftBorder"
    href="/docs/"
    target="_blank"
    title="Open the documentation in a new window"
  >
    Docs
    <i
      class="open-in-new"
    />
  </a>
</div>
`;

exports[`app/MenuButtons <Publish> matches the snapshot for the menu buttons and the opened panel for an already uploaded profile 2`] = `
<div
  data-testid="MenuButtonsPublish-container"
>
  <form
    class="menuButtonsPublishContent"
  >
    <div
      class="menuButtonsPublishIcon"
    />
    <h1
      class="menuButtonsPublishTitle"
    >
      Re-upload Performance Profile
    </h1>
    <p
      class="menuButtonsPublishInfoDescription"
    >
      Upload your profile and make it accessible to anyone with the link.
       By default, your personal data is removed.
    </p>
    <h3>
      Include additional data that may be identifiable
    </h3>
    <div
      class="menuButtonsPublishDataChoices"
    >
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeHiddenThreads"
          type="checkbox"
        />
        Include hidden threads
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeFullTimeRange"
          type="checkbox"
        />
        Include hidden time range
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeScreenshots"
          type="checkbox"
        />
        Include screenshots
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeUrls"
          type="checkbox"
        />
        Include resource URLs and paths
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeExtension"
          type="checkbox"
        />
        Include extension information
      </label>
    </div>
    <div
      class="menuButtonsPublishButtons"
    >
      <button
        class="photon-button menuButtonsPublishButton menuButtonsPublishButtonsDownload menuButtonsPublishButtonDisabled"
        type="button"
      >
        Compressing…
      </button>
      <button
        class="photon-button photon-button-primary menuButtonsPublishButton menuButtonsPublishButtonsUpload"
        type="submit"
      >
        <span
          class="menuButtonsPublishButtonsSvg menuButtonsPublishButtonsSvgUpload"
        />
        Upload
      </button>
    </div>
  </form>
</div>
`;

exports[`app/MenuButtons <Publish> matches the snapshot for the opened panel for a nightly profile 1`] = `
<div
  data-testid="MenuButtonsPublish-container"
>
  <form
    class="menuButtonsPublishContent"
  >
    <div
      class="menuButtonsPublishIcon"
    />
    <h1
      class="menuButtonsPublishTitle"
    >
      Share Performance Profile
    </h1>
    <p
      class="menuButtonsPublishInfoDescription"
    >
      Upload your profile and make it accessible to anyone with the link.
       This profile is from Firefox Nightly, so by default all information is included.
    </p>
    <h3>
      Include additional data that may be identifiable
    </h3>
    <div
      class="menuButtonsPublishDataChoices"
    >
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          checked=""
          class="photon-checkbox photon-checkbox-default"
          name="includeHiddenThreads"
          type="checkbox"
        />
        Include hidden threads
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          checked=""
          class="photon-checkbox photon-checkbox-default"
          name="includeFullTimeRange"
          type="checkbox"
        />
        Include hidden time range
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          checked=""
          class="photon-checkbox photon-checkbox-default"
          name="includeScreenshots"
          type="checkbox"
        />
        Include screenshots
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          checked=""
          class="photon-checkbox photon-checkbox-default"
          name="includeUrls"
          type="checkbox"
        />
        Include resource URLs and paths
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          checked=""
          class="photon-checkbox photon-checkbox-default"
          name="includeExtension"
          type="checkbox"
        />
        Include extension information
      </label>
    </div>
    <div
      class="menuButtonsPublishButtons"
    >
      <button
        class="photon-button menuButtonsPublishButton menuButtonsPublishButtonsDownload menuButtonsPublishButtonDisabled"
        type="button"
      >
        Compressing…
      </button>
      <button
        class="photon-button photon-button-primary menuButtonsPublishButton menuButtonsPublishButtonsUpload"
        type="submit"
      >
        <span
          class="menuButtonsPublishButtonsSvg menuButtonsPublishButtonsSvgUpload"
        />
        Upload
      </button>
    </div>
  </form>
</div>
`;

exports[`app/MenuButtons <Publish> matches the snapshot for the opened panel for a release profile 1`] = `
<div
  data-testid="MenuButtonsPublish-container"
>
  <form
    class="menuButtonsPublishContent"
  >
    <div
      class="menuButtonsPublishIcon"
    />
    <h1
      class="menuButtonsPublishTitle"
    >
      Share Performance Profile
    </h1>
    <p
      class="menuButtonsPublishInfoDescription"
    >
      Upload your profile and make it accessible to anyone with the link.
       By default, your personal data is removed.
    </p>
    <h3>
      Include additional data that may be identifiable
    </h3>
    <div
      class="menuButtonsPublishDataChoices"
    >
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeHiddenThreads"
          type="checkbox"
        />
        Include hidden threads
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeFullTimeRange"
          type="checkbox"
        />
        Include hidden time range
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeScreenshots"
          type="checkbox"
        />
        Include screenshots
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeUrls"
          type="checkbox"
        />
        Include resource URLs and paths
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeExtension"
          type="checkbox"
        />
        Include extension information
      </label>
    </div>
    <div
      class="menuButtonsPublishButtons"
    >
      <button
        class="photon-button menuButtonsPublishButton menuButtonsPublishButtonsDownload menuButtonsPublishButtonDisabled"
        type="button"
      >
        Compressing…
      </button>
      <button
        class="photon-button photon-button-primary menuButtonsPublishButton menuButtonsPublishButtonsUpload"
        type="submit"
      >
        <span
          class="menuButtonsPublishButtonsSvg menuButtonsPublishButtonsSvgUpload"
        />
        Upload
      </button>
    </div>
  </form>
</div>
`;
